vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs

```diff
@@ -411,25 +416,64 @@ module internal Extensions =
                 return! this.ParseDocument(document, options, sourceText.ToString())
             }
 
-        member this.ParseAndCheckDocument(filePath: string, textVersionHash: int, sourceText: string, options: FSharpProjectOptions) : Async<(Ast.ParsedInput * FSharpCheckFileResults) option> =
-            async {
-                let! parseResults, checkFileAnswer = this.ParseAndCheckFileInProject(filePath, textVersionHash, sourceText, options)
-                return
-                    match parseResults.ParseTree, checkFileAnswer with
-                    | _, FSharpCheckFileAnswer.Aborted 
-                    | None, _ -> None
-                    | Some parsedInput, FSharpCheckFileAnswer.Succeeded checkResults -> Some (parsedInput, checkResults)
-            }
+        member this.ParseAndCheckDocument(filePath: string, textVersionHash: int, sourceText: string, options: FSharpProjectOptions, allowStaleResults: bool) : Async<(FSharpParseFileResults * Ast.ParsedInput * FSharpCheckFileResults) option> =
+            let parseAndCheckFile =
+                async {
+                    let! parseResults, checkFileAnswer = this.ParseAndCheckFileInProject(filePath, textVersionHash, sourceText, options)
+                    return
+                        match checkFileAnswer with
+                        | FSharpCheckFileAnswer.Aborted -> 
+                            None
+                        | FSharpCheckFileAnswer.Succeeded(checkFileResults) ->
+                            Some (parseResults, checkFileResults)
+                }
+
+            let tryGetFreshResultsWithTimeout(timeout: int) : Async<CheckResults> =
+                async {
+                    try
+                        let! worker = Async.StartChild(parseAndCheckFile, timeout)
+                        let! result = worker 
+                        return Ready result
+                    with :? TimeoutException ->
+                        return StillRunning parseAndCheckFile
+                }
+
+            let bindParsedInput(results: (FSharpParseFileResults * FSharpCheckFileResults) option) =
+                match results with
+                | Some(parseResults, checkResults) ->
+                    match parseResults.ParseTree with
+                    | Some parsedInput -> Some (parseResults, parsedInput, checkResults)
+                    | None -> None
+                | None -> None
 
-        member this.ParseAndCheckDocument(document: Document, options: FSharpProjectOptions, ?sourceText: SourceText) : Async<(Ast.ParsedInput * FSharpCheckFileResults) option> =
+            if allowStaleResults then
+                async {
+                    let! freshResults = tryGetFreshResultsWithTimeout 1000
```

Comment:
Please make this timeout settable at least via an environment variable, and (much preferably) via a user-facing configuration option.